### DISCUSSION TITLE
useResponseCache: How to invalidate cache with apollo federation

### DISCUSSION THREAD

**Discussion by kanopa** [dotansimha/envelop#1480] [Q&A] [answered]
I have two services that communicate with each other through a gateway using apollo federation, search and main backend with data
example query search where name resolve through federation
```graphql
query Search($first: Int, $query: String) {
  search(query: $query) {
    example(first: $first) {
      nodes {
        id
        name
      }
    }
  }
}
```
When making a request, a request for one service saves only id and _typename, and the second service saves all other requested fields without id and _typename
<img width="975" alt="image" src="https://user-images.githubusercontent.com/42976537/184717117-f0910637-185f-4803-be5f-1b9f72c60c31.png">
After mutation only one of records is invalidating (record with id and _typename)
I did a little research on this, data with _typename and id on federation requests exists in `onExecuteParams.args.variableValues`
`representations: [{ __typename: 'Post', id: '2838935f-751d-4497-b48a-5c747a6c75e9' }]`

How to properly invalidate cache using apollo federation?

**Comment by kanopa** [comment, marked as answer]
I guess we need sth like this and call it on onExecuteDone
```js

const processFederationResult = (processedResult) => {
                // if _entities exists in data it is federation response
                if ('_entities' in processedResult?.data) {
                    // need to add all representations to identifier for invalidation federation responses
                    onExecuteParams.args.variableValues.representations?.forEach(({ __typename: typename, id }) => {
                        identifier.set(`${typename}:${id}`, { typename, id });
                    })
                }
            }
```

**Reply by n1ru4l** [reply]
@kanopa do you think you can create a pull request on this repo that adds such an example to the examples folder? I think this will be a valuable resource for everyone using federation!

**Reply by kanopa** [reply]
> @kanopa do you think you can create a pull request on this repo that adds such an example to the examples folder? I think this will be a valuable resource for everyone using federation!

I had to do it using yarn patch and edited useResponseCache straight, maybe we can integrated this with useResponseCache?

**Reply by n1ru4l** [reply]
@kanopa Would this help you solve this issue in userland? https://github.com/n1ru4l/envelop/pull/1517

**Reply by n1ru4l** [reply]
Hey @kanopa, did you have time to look into this PR?

**Reply by n1ru4l** [reply]
Hey @kanopa, any update?

**Reply by n1ru4l** [reply]
@kanopa I will close the PR and mark this discussion as solved, as there seems no need for this feature anymore.
